homeassistant/components/motioneye/__init__.py

```diff
@@ -109,13 +147,89 @@ def _add_camera(
 ) -> None:
     """Add a motionEye camera to hass."""
 
-    device_registry.async_get_or_create(
+    def _is_recognized_web_hook(url: str) -> bool:
+        """Determine whether this integration set a web hook."""
+        return f"{WEB_HOOK_SENTINEL_KEY}={WEB_HOOK_SENTINEL_VALUE}" in url
+
+    def _set_webhook(
+        url: str,
+        key_url: str,
+        key_method: str,
+        key_enabled: str,
+        camera: dict[str, Any],
+    ) -> bool:
+        """Set a web hook."""
+        if (
+            entry.options.get(
+                CONF_WEBHOOK_SET_OVERWRITE,
+                DEFAULT_WEBHOOK_SET_OVERWRITE,
+            )
+            or not camera.get(key_url)
+            or _is_recognized_web_hook(camera[key_url])
+        ) and (
+            not camera.get(key_enabled, False)
+            or camera.get(key_method) != KEY_HTTP_METHOD_POST_JSON
+            or camera.get(key_url) != url
+        ):
+            camera[key_enabled] = True
+            camera[key_method] = KEY_HTTP_METHOD_POST_JSON
+            camera[key_url] = url
+            return True
+        return False
+
+    def _build_url(
+        device: dr.DeviceEntry, base: str, event_type: str, keys: list[str]
+    ) -> str:
+        """Build a motionEye webhook URL."""
+
+        return urljoin(
+            base,
+            "?"
+            + urlencode(
+                {
+                    **{k: KEY_WEB_HOOK_CONVERSION_SPECIFIERS[k] for k in sorted(keys)},
+                    WEB_HOOK_SENTINEL_KEY: WEB_HOOK_SENTINEL_VALUE,
+                    ATTR_EVENT_TYPE: event_type,
+                    ATTR_DEVICE_ID: device.id,
+                },
+                safe="%{}",
+            ),
+        )
+
+    device = device_registry.async_get_or_create(
         config_entry_id=entry.entry_id,
         identifiers={device_identifier},
         manufacturer=MOTIONEYE_MANUFACTURER,
         model=MOTIONEYE_MANUFACTURER,
         name=camera[KEY_NAME],
     )
+    if entry.options.get(CONF_WEBHOOK_SET, DEFAULT_WEBHOOK_SET):
+        url = async_generate_url(hass, entry.data[CONF_WEBHOOK_ID])
```

Comment:
This will generate an external URL for the webhook, if available.

Should we check for active HA cloud and use that as well?

`register_webhook` in Netatmo might be a good example of this.

Reply:
Had a shot. Please take a look.

Note: I have no way to test this beyond unittests.